src/ray/gcs/gcs_server/gcs_health_check_manager.cc

```diff
@@ -118,6 +151,9 @@ void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
                 delete this;
```

Comment:
The best way, IMO, is to store `weak_ptr` in map, and capture with shared pointer.
I could followup with another PR --- `delete this` is not incorrect, but error prune.

Reply:
what about keeping shared_ptr in map, and in capture, and `--health_check_remaining_;` remove self from the map?

Reply:
That's what I think, place weak pointer vs shared pointer in the map has no difference.

Reply:
I don't like the idea of the map keeping weak ptrs, sounds like prone to mem leak (map never erased) to me.

Reply:
Using weak pointer is better than shared pointer here (though they're of the same effect), the reason is, weak pointer doesn't account for ref count.

Reply:
I think it's mostly about ownership models. I feel the map should own the context objs so it should take a refcount. The lambdas should take one as well, and after the lambda erases it from map and the lambda itself goes out of scope (? need more thoughts on this) the ctx obj is released.

anyway since we are not changing this in this PR - leave for future discussions.

Reply:
One common usage for weak cache for zero-copy purpose, is to store weak reference at map, so key-value pair could be evicted at customized deleter for shared pointer.

Reply:
ok. one caveat is that the dtor may be called at surprising places, e.g. grpc's threads. and it may need a mutex to properly erase from the map, which defeats the purpose of any optimizations.

Reply:
specific to this callsite, (I assume) the dtor happens on io_context's code so it should be in the right thread.